pkg/assets/stage.go

```diff
@@ -18,53 +18,67 @@ package assets
 
 import (
 	"compress/gzip"
+	"errors"
 	"fmt"
 	"io"
 	"os"
 	"os/exec"
 	"path/filepath"
+	"syscall"
+
+	"github.com/k0sproject/k0s/internal/pkg/file"
+	"github.com/k0sproject/k0s/pkg/constant"
 
 	"github.com/sirupsen/logrus"
 )
 
-// EmbeddedBinaryNeedsUpdate returns true if the provided embedded binary file should
-// be updated. This determination is based on the modification times and file sizes of both
-// the provided executable and the embedded executable. It is expected that the embedded binary
-// modification times should match the main `k0s` executable.
-func EmbeddedBinaryNeedsUpdate(exinfo os.FileInfo, embeddedBinaryPath string, size int64) bool {
-	if pathinfo, err := os.Stat(embeddedBinaryPath); err == nil {
-		return !exinfo.ModTime().Equal(pathinfo.ModTime()) || pathinfo.Size() != size
+// Stages the embedded executable with the given name into dir. If the
+// executable is not embedded in the k0s executable, this function first checks
+// if an executable exists at the desired path. If not, it falls back to a PATH
+// lookup. Returns the path to the executable, even if an error occurs.
+func StageExecutable(dir, name string) (string, error) {
+	// Always returning the path, even under error conditions, is kind of a hack
+	// to work around the "running executable" problem on Windows.
+
+	executableName := name + constant.ExecutableSuffix
+	path := filepath.Join(dir, executableName)
+	err := stage(executableName, path, 0750)
```

Comment:
0750 should probably be a constant in pkg/constant/constant.go.
Anyway that's just a nit.